System.Net.Sockets tests crashed in CI on Windows

http://dotnet-ci.cloudapp.net/job/dotnet_corefx_windows_release_prtest/5725/console
```
Discovering: System.Net.Sockets.Tests
  Discovered:  System.Net.Sockets.Tests
  Starting:    System.Net.Sockets.Tests
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00111\lib\tests.targets(129,5): warning MSB3073: The command "CoreRun.exe xunit.console.netcore.exe System.Net.Sockets.Tests.dll  -xml testResults.xml -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing " exited with code -1073741819. [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\tests\FunctionalTests\System.Net.Sockets.Tests.csproj]
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00111\lib\tests.targets(136,5): error : One or more tests failed while running tests from 'System.Net.Sockets.Tests' please check log for details! [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\tests\FunctionalTests\System.Net.Sockets.Tests.csproj]
```

Thanks @weshaggard but that's the log that @stephentoub attached as well. 

This seems to be the only relevant portion from the log and it's not telling me why the tests failed:
```
     System.Net.Sockets.AsyncServer.Tests  Total: 354, Errors: 0, Failed: 0, Skipped: 0, Time: 66.825s
  System.Net.Sockets.APM.Performance.Tests -> d:\j\workspace\dotnet_corefx_windows_release_prtest\bin\Windows_NT.AnyCPU.Release\System.Net.Sockets.APM.Performance.Tests\System.Net.Sockets.APM.Performance.Tests.dll
  System.Net.Sockets.Tests -> d:\j\workspace\dotnet_corefx_windows_release_prtest\bin\Windows_NT.AnyCPU.Release\System.Net.Sockets.Tests\System.Net.Sockets.Tests.dll
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.
  
  Discovering: System.Net.Sockets.APM.Performance.Tests
  Discovered:  System.Net.Sockets.APM.Performance.Tests
  Info:        System.Net.Sockets.APM.Performance.Tests has no tests to run
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.
  
  Discovering: System.Net.Sockets.Tests
  Discovered:  System.Net.Sockets.Tests
  Starting:    System.Net.Sockets.Tests
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00111\lib\tests.targets(129,5): warning MSB3073: The command "CoreRun.exe xunit.console.netcore.exe System.Net.Sockets.Tests.dll  -xml testResults.xml -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing " exited with code -1073741819. [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\tests\FunctionalTests\System.Net.Sockets.Tests.csproj]
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00111\lib\tests.targets(136,5): error : One or more tests failed while running tests from 'System.Net.Sockets.Tests' please check log for details! [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\tests\FunctionalTests\System.Net.Sockets.Tests.csproj]
  System.Net.Sockets.Async.Performance.Tests -> d:\j\workspace\dotnet_corefx_windows_release_prtest\bin\Windows_NT.AnyCPU.Release\System.Net.Sockets.Async.Performance.Tests\System.Net.Sockets.Async.Performance.Tests.dll
  System.Reflection.Metadata.Tests -> d:\j\workspace\dotnet_corefx_windows_release_prtest\bin\Windows_NT.AnyCPU.Release\System.Reflection.Metadata.Tests\System.Reflection.Metadata.Tests.dll
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.
```

This issue seems to be a duplicate of #4037.
The investigation is currently (partially) _blocked_ by #4152.

Thanks @weshaggard for suggesting offline that this could be caused by an Assert. Given that the issue is most likely timing, I'll try to repro on a low-resource machine until we can get at least part of #4152 ready.

--- a/src/Common/src/System/Net/SafeCloseSocket.cs
+++ b/src/Common/src/System/Net/SafeCloseSocket.cs
@@ -52,9 +52,10 @@ namespace System.Net.Sockets
             try
             {
                 // The inner socket can be closed by CloseAsIs and when SafeHandle runs ReleaseHandle.
-                if (_innerSocket != null)
+                InnerSafeCloseSocket innerSocket = Volatile.Read(ref _innerSocket);
+                if (innerSocket != null)
                 {
-                    _innerSocket.AddRef();
+                    innerSocket.AddRef();
                 }
             }
             catch (Exception e)
@@ -68,9 +69,10 @@ namespace System.Net.Sockets
             try
             {
                 // The inner socket can be closed by CloseAsIs and when SafeHandle runs ReleaseHandle.
-                if (_innerSocket != null)
+                InnerSafeCloseSocket innerSocket = Volatile.Read(ref _innerSocket);
+                if (innerSocket != null)
                 {
-                    _innerSocket.Release();
+                    innerSocket.Release();
                 }
             }
             catch (Exception e)

